### ISSUE TITLE
Optional semicolons in scripts

### ISSUE THREAD

**Issue by edmundito** [adventuregamestudio/ags#2017]
**Describe the problem**
Semicolons in scripting  now days are mostly aesthetic. They are good when writing multiple statements in a single line, but only some people take advantage of this. A lot of more modern languages either avoid them or make them optional

**Suggested change**
Make them optional. If they are there, no problem. If they are not, new line is the separator for statements. The demo game and templates will no longer include them, and moving forward they are not encouraged.

**Comment by ericoporto** [comment]
I don't have good feelings on this, I see a lot of badly formated code in the forums from people that are just starting up and having no semicolons would make reading their code even harder...

Semicolons tells in unambiguous way when the specifying exactly where the "instructions" end.

**Comment by ivan-mogilko** [comment]
> Semicolons in scripting now days are mostly aesthetic. They are good when writing multiple statements in a single line, but only some people take advantage of this.

I feel like there's an erroneous assumption here. Semicolons are used not only to have several expressions in a line, they are also used to have several lines per expression.

Consider this code:
```
int a = 1 + 2
           + 3 + 4
           + 5 + 6;
```
This code will have different meaning depending on which syntax rule is active.

In fact the code you posted in another ticket:
```
player
  .Say("Hang on...")
  .Animate(4, 1)
  .Say("This door is locked?!")
```
This code will also have different meaning depending on whether compiler is looking for semicolon to end a statement or not.

**Comment by ivan-mogilko** [comment]
> Semicolons in scripting now days are mostly aesthetic. They are good when writing multiple statements in a single line, but only some people take advantage of this

Semicolons are also used when you are having multiple lines per statements: this way you clearly know where the statement ends.

If semicolon is omitted, the compiler will have to read not until the current part of expression ends, but also further, until it sees the clear beginning of a new statement.

Consider code:
```
int a = 1 + 2
        + 3 + 4
        + 5 + 6
int b;
```
Compiler will have to keep reading until it meets `int b` and then it will realize that the previous assignment ended.

The question that I have is: do we have any syntactical construct in ags script (or may we get one in a future) that may have dubious meaning in such context. that is - which may not be unambiguously identified as whether a beginning of a new statement or a continuation of previous one?

**Comment by fernewelten** [comment]
> The question that I have is: do we have any syntactical construct in ags script (or may we get one in a future) that may have dubious meaning in such context. that is - which may not be unambiguously identified as whether a beginning of a new statement or a continuation of previous one?

```
function Weaken(int Stamina)
{
    return
        Stamina--
}
```
can either return or return `Stamina--`.

**Comment by fernewelten** [comment]
> A lot of more modern languages either avoid them or make them optional

Can you please expand on “a lot“? I know PHP does that, but AFAIK the whole group of C-style languages does not. Java does not. JavaScript does not behind-the-scenes, either,  AFAIK, but there's some kludgy wrangling stuck on top that allows some semicolons to be implied.

Personally, I don't really like the idea of implied statement ends. It doesn't save all that much typing.

At the moment, AGS is mostly modelled after C#, and C# requires semicolons.
